### PR TITLE
MAM4xx  port of nucleate_ice

### PR DESCRIPTION
**Pull request by odiazib**
MAM4xx  port of `nucleate_ice`

This is the MAM4xx port of subroutines `nucleati`, `hetero`, `hf` from [nucleate_ice.F90](https://github.com/eagles-project/e3sm_mam4_refactor/blob/refactor-maint-2.0/components/eam/src/physics/cam/nucleate_ice.F90) and subroutine `nucleate_ice_cam`, which is the upper level interface that calls `nucleati`, from [nucleate_ice_cam.F90](https://github.com/eagles-project/e3sm_mam4_refactor/blob/refactor-maint-2.0/components/eam/src/physics/cam/nucleate_ice_cam.F90). The subroutine `nucleate_ice_cam` outputs six variables that mam4xx stores in diagnostics. The variables are:
`naai`(`num_act_aerosol_ice_nucle`), `naai_hom`(`num_act_aerosol_ice_nucle_hom`), `nihf`(`icenuc_num_hetfrz`,), `niimm`(`icenuc_num_immfrz`), `nidep`(`icenuc_num_depnuc`), `nimey`(`icenuc_num_meydep`) where the name in parenthesis is the variable name in mam4xx-diagnostics.

Two parameters in the above subroutines (`subgrid` and `so4_sz_thresh_icenuc`) are initialized in `nucleate_ice_cam_readnl` and `nucleati_init`. We initialize these parameters in the nucleate_ice class’ constructor. Furthermore, `nucleate_ice` and `nucleate_ice_cam` use function calls from [wv_saturation.F90](https://github.com/eagles-project/e3sm_mam4_refactor/blob/refactor-maint-2.0/components/eam/src/physics/cam/wv_saturation.F90) and [wv_sat_methods.F90](https://github.com/eagles-project/e3sm_mam4_refactor/blob/refactor-maint-2.0/components/eam/src/physics/cam/wv_sat_methods.F90) to compute the saturation water vapor pressure for water and ice. We port the functions required by nucleate_ice to the separate header file [wv_sat_methods.hpp](https://github.com/eagles-project/mam4xx/blob/collab/nucleate_ice/src/mam4xx/wv_sat_methods.hpp).

@mingxuanwupnnl generated a set  of validation tests for `nucleati`, `hetero`, `hf`, and `nucleate_ice_cam` using e3sm. We create validation tests in mam4xx and verify that all tests produced small relative tolerances (Biggest passing tolerances is 2e-4 in nucleate_ice_cam validation test). In addition, we create a driver that produces skywalker files for nucleate_ice in [mam4_standalone_validation](https://github.com/eagles-project/mam4_standalone_validation). Using this driver, we generated an additional validation test varying across three different variables. We find that for these tests `mam4xx::nucleate_ice::nucleati()` matches the values from the Fortran subroutine bit-for-bit.

We do want to draw attention to a few places with odd behavior that we leave unchanged so as to match the Fortran code. First, we note that two of the above diagnostic variables are hard-wired to value of zero, i.e., `nimey` in [L214--MAM4 Refactor](https://github.com/eagles-project/e3sm_mam4_refactor/blob/8570bb3cbd2a0abf65ec86a4bb2fe2f69e09248b/components/eam/src/physics/cam/nucleate_ice.F90#L214) and `nidep` in [L275L214--MAM4 Refactor](https://github.com/eagles-project/e3sm_mam4_refactor/blob/8570bb3cbd2a0abf65ec86a4bb2fe2f69e09248b/components/eam/src/physics/cam/nucleate_ice.F90#L275). In the case of L214, `nimey` is subsequently added to `nuci`, after which `nuci` is fed to an if statement, to see if it is within a given interval and if true, `nuci` is also set to zero.

### CODE REVIEW

**Comment by jeff-cohere** [comment]
Hi guys. Can you push one more commit with `[rebuild-haero]` in the commit message? The PNNL CI pipeline is failing because of a missing field in `Haero::Constants`, so we need to make sure Haero is getting rebuilt:

```
...
/people/svceagles/gitlab/4200/haero_Release_single/src/mam4xx/wv_sat_methods.hpp(18): error: class "haero::Constants" has no member "boil_pt_h2o"
...
```

File: src/validation/nucleate_ice/CMakeLists.txt [hunk 1]
```diff
@@ -25,14 +25,15 @@ endforeach()
 
 set(TEST_LIST
     nucleate_ice_cam_calc_merged # passing tolerance = 1.5e-5
-    hetero_merged
+    # hetero_merged
```
Comment (jeff-cohere) [inline src/validation/nucleate_ice/CMakeLists.txt]: are these tests just commented out for the moment, to be reintroduced in a subsequent PR?

File: src/validation/nucleate_ice/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    hf_merged
+    nucleati_merged # passing tolerance = 3e-9
+    nucleate_ice # produced by mam4_standalone_validation
+    )
+set(DEFAULT_TOL 1e-9)
+set(ERROR_THRESHOLDS 1.5e-5 ${DEFAULT_TOL} ${DEFAULT_TOL} 3e-9 ${DEFAULT_TOL})
+
+# Run the driver in several configurations to produce datasets.
+
+foreach(input tol IN ZIP_LISTS TEST_LIST ERROR_THRESHOLDS)
```
Comment (jeff-cohere) [inline src/validation/nucleate_ice/CMakeLists.txt:39]: Aha, `ZIP_LISTS`. Neat!

**Review by jeff-cohere — APPROVED** [review]
This looks good to me. Nice work, fellas. Is there anything you want to discuss? I know there are a few oddities that you encountered. Do we need to create any issues to bookmark any of these, and/or discuss resolutions?

EDIT: looks like another unused variable is spoiling one of the builds. I'm not sure why this doesn't show up in the others.

**Comment by jeff-cohere** [comment]
Huh. Now it's taking forever in our offending configuration. This configuration seems to have real trouble when we build in parallel on the GitHub Actions test runners. It's a pity we can't reproduce this on our own hardware.

**Comment by jeff-cohere** [comment]
🤔 it may be easiest just to use `make` on line 91 for the moment, and see whether that has a huge effect on our build/test time.

**Comment by mjs271** [comment]
> 🤔 it may be easiest just to use `make` on line 91 for the moment, and see whether that has a huge effect on our build/test time.

... yep. Was trying to get clever. Ended up outsmarting myself 🙃

**Comment by jeff-cohere** [comment]
> ... yep. Was trying to get clever and thought it would be easy. Ended up outsmarting myself 🙃

I certainly know that feeling! 😄

**Comment by jeff-cohere** [comment]
Great! Can you re-add the `nucleate_ice()` unit tests?

**Comment by jeff-cohere** [comment]
Hey @pressel , @overfelt @mingxuanwupnnl @singhbalwinder . Can you take a look at this PR and give your feedback?

File: src/mam4xx/wv_sat_methods.hpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  es = haero::min(es, p);
+
+} // wv_sat_qsat_water
+
+KOKKOS_INLINE_FUNCTION
+Real svp_ice(const Real temperature) { return GoffGratch_svp_ice(temperature); }
+
+} // namespace wv_sat_methods
+} // namespace mam4
+
```
Comment (pressel) [inline src/mam4xx/wv_sat_methods.hpp:136]: Note that most if not all of this is also implemented as part of SCREAM, we should check for consistency if we are implementing it again.

Comment (odiazib) [reply src/mam4xx/wv_sat_methods.hpp:136]: if these functions are implemented in SCREAM, how should we invoke them in mam4xx ?

Comment (jeff-cohere) [reply src/mam4xx/wv_sat_methods.hpp:136]: For now, I think we should make a note of them. We don't have a mechanism to call external code like this, but if we wanted to support that, we could probably figure out how to add some smarts to the build system to allow it.

**Review by mingxuanwupnnl — APPROVED** [review]
The C++ porting of nucleate_ice/nucleate_ice_cam.F90 looks good to me. I don't have other comments.

**Review by pressel — APPROVED** [review]
This looks great to me. Thanks for all of you work on this @odiazib and @mjs271.

File: src/mam4xx/wv_sat_methods.hpp (hunk 3 above)
Comment (pressel) [reply src/mam4xx/wv_sat_methods.hpp:136]: Agreed @jeff-cohere. I just wanted to make a note for the record.

**Review by mjs271 — APPROVED** [review]

**Pull request merged**